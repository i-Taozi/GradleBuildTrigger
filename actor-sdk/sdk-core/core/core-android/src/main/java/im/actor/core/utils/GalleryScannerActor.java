package im.actor.core.utils;

import android.content.Context;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Environment;
import android.provider.MediaStore;

import java.util.ArrayList;
import java.util.HashMap;
import im.actor.core.viewmodel.GalleryVM;
import im.actor.runtime.Log;
import im.actor.runtime.actors.Actor;
import im.actor.runtime.collections.SparseArray;

public class GalleryScannerActor extends Actor {

    public static final String TAG = "GALLERY_SCANNER";
    public static final int CHECK_NEW_DELAY = 1000;
    public static final int SCAN_DELAY = 100;
    Context context;
    Uri uri;
    Cursor cursor;
    int offset = 0;
    int column_index_data, column_index_folder_name, column_index_date, column_index_id, column_index_folder_id;

    ArrayList<String> listOfAllImages = new ArrayList<>();
    ArrayList<String> newMedia = new ArrayList<>();

    String absolutePathOfImage = null;
    GalleryVM galleryVM;

    private static final int SCAN_COUNT = 10;

    private boolean visible = false;
    private boolean scanned = false;
    private String[] projection;
    private boolean inited = false;
    private Bitmap bitmap;

    public GalleryScannerActor(Context context, GalleryVM galleryVM) {
        this.context = context;
        this.galleryVM = galleryVM;
    }

    @Override
    public void preStart() {
        uri = android.provider.MediaStore.Images.Media.EXTERNAL_CONTENT_URI;
    }

    private void initScan() {
        Log.d(TAG, "init scan");
        projection = new String[]{MediaStore.Images.Media.DATA,
                MediaStore.Images.Media._ID,
                MediaStore.Images.Media.BUCKET_ID,
                MediaStore.Images.Media.BUCKET_DISPLAY_NAME,
                MediaStore.Images.Media.DATE_TAKEN};

        cursor = getQuery();

        if (cursor != null && cursor.getCount() > 0) {
            column_index_data = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA);
            column_index_id = cursor.getColumnIndexOrThrow(MediaStore.Images.Media._ID);
            column_index_folder_id = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.BUCKET_ID);
            column_index_date = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATE_TAKEN);
            column_index_folder_name = cursor
                    .getColumnIndexOrThrow(MediaStore.Images.Media.BUCKET_DISPLAY_NAME);


            Log.d(TAG, "init scan ok - starting scan iterations");
            inited = true;
            self().send(new Scan());
        } else {
            Log.d(TAG, "init scan - no media, let's check it in " + CHECK_NEW_DELAY);
            if (visible) {
                schedule(new InitScan(), CHECK_NEW_DELAY);
            }
        }

    }

    private Cursor getQuery() {
        try {
            return context.getContentResolver().query(uri, projection, null,
                    null, MediaStore.Images.Media.DATE_TAKEN + " DESC");
        } catch (Exception e) {
            return null;
        }
    }

    private void scan() {

        Log.d(TAG, "scan");

        for (int i = 0; i <= SCAN_COUNT && cursor.moveToNext(); i++) {
            absolutePathOfImage = cursor.getString(column_index_data);
            listOfAllImages.add(absolutePathOfImage);
        }

        offset += SCAN_COUNT;
        scanned = offset >= cursor.getCount();

        //
        // Update VM after 1st batch, every 100 photos, when scanned to end
        //
        if (offset == SCAN_COUNT || offset % 100 == 0 || scanned) {
            Log.d(TAG, "scan - update vm, offset - " + offset);
            galleryVM.getGalleryMediaPath().change(new ArrayList<String>(listOfAllImages));
        }

        if (visible) {
            if (!scanned) {
                schedule(new Scan(), SCAN_DELAY);
            } else {
                schedule(new CheckNew(), CHECK_NEW_DELAY);
            }
        }
    }

    private void checkNew() {
        Log.d(TAG, "checkNew");
        cursor = getQuery();
        newMedia.clear();
        boolean firstCycle = true;
        while (cursor != null && cursor.moveToNext()) {
            absolutePathOfImage = cursor.getString(column_index_data);
            if (!listOfAllImages.contains(absolutePathOfImage)) {
                if (firstCycle) {
                    firstCycle = false;
                    bitmap = BitmapFactory.decodeFile(absolutePathOfImage);
                    if (bitmap != null) {
                        bitmap.recycle();
                        newMedia.add(absolutePathOfImage);
                        Log.d(TAG, "checkNew - new media");

                    } else {
                        Log.d(TAG, "checkNew - new media writing, breaking for wait");

                        break;
                    }
                } else {
                    newMedia.add(absolutePathOfImage);
                }

            } else {
                Log.d(TAG, "checkNew - found old media, break");

                break;
            }
        }
        if (newMedia.size() > 0) {
            listOfAllImages.addAll(0, newMedia);
            galleryVM.getGalleryMediaPath().change(new ArrayList<>(listOfAllImages));
            Log.d(TAG, "checkNew - new media add - " + newMedia.size());
        }
        if (visible) {
            Log.d(TAG, "checkNew - visible, schedule next check in " + CHECK_NEW_DELAY);
            schedule(new CheckNew(), CHECK_NEW_DELAY);
        }
    }

    @Override
    public void onReceive(Object message) {
        if (message instanceof Scan) {
            scan();
        } else if (message instanceof InitScan) {
            initScan();
        } else if (message instanceof CheckNew) {
            checkNew();
        } else if (message instanceof Show) {
            visible = true;
            if (scanned) {
                self().send(new CheckNew());
            } else {
                if (!inited) {
                    self().send(new InitScan());
                }
            }
        } else if (message instanceof Hide) {
            visible = false;
        }
    }

    private static class Scan {

    }

    private static class InitScan {

    }

    private static class CheckNew {

    }

    public static class Show {

    }

    public static class Hide {

    }
}
