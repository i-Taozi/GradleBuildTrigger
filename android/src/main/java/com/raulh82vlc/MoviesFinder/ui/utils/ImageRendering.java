package com.raulh82vlc.MoviesFinder.ui.utils;

import android.widget.ImageView;

/**
 * Image Rendering contract
 */
public interface ImageRendering {
    /**
     * load an image
     *
     * @param urlPath path of the image from the API
     * @param img     ImageView where to render
     */
    void loadImage(String urlPath, ImageView img);
}
