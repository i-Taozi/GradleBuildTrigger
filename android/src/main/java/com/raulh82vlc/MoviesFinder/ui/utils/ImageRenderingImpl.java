/*
 * Copyright (C) 2016 Raul Hernandez Lopez
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.raulh82vlc.MoviesFinder.ui.utils;

import android.app.Activity;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.RequestManager;
import com.bumptech.glide.load.engine.DiskCacheStrategy;

import javax.inject.Inject;

/**
 * Image Rendering
 *
 * @author Raul Hernandez Lopez
 */
public class ImageRenderingImpl implements ImageRendering {

    private RequestManager mGlideInstance;

    @Inject
    public ImageRenderingImpl(Activity context) {
        // this might not lead to a null pointer exception but just in case re-initialisation
        if (mGlideInstance == null) {
            synchronized (this) {
                if (mGlideInstance == null) {
                    mGlideInstance = Glide.with(context);
                }
            }
        }
    }

    @Override
    public void loadImage(String urlPath, ImageView img) {
        mGlideInstance
                .load(urlPath)
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .fitCenter()
                .into(img);
    }
}
