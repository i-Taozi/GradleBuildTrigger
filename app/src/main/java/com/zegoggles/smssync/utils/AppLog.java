package com.zegoggles.smssync.utils;

import android.app.Dialog;
import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;

import android.text.format.DateFormat;
import android.util.Log;
import android.widget.ScrollView;
import android.widget.TextView;
import com.zegoggles.smssync.App;
import com.zegoggles.smssync.R;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.LineNumberReader;
import java.io.PrintWriter;
import java.util.Date;

import static com.zegoggles.smssync.App.LOCAL_LOGV;
import static com.zegoggles.smssync.App.TAG;

public class AppLog {
    // keep max 32k worth of logs
    private static final int MAX_SIZE = 32 * 1024;
    private static final int ID = 1;

    private @Nullable PrintWriter writer;
    private String dateFormat;

    public AppLog(Context context) {
        char[] format;
        try {
            format = DateFormat.getDateFormatOrder(context);
        } catch (IllegalArgumentException e) {
            format = new char[] { 'd' };
        }
        for (char c : format) {
            if (c == 'M') {
                dateFormat = "MM-dd kk:mm";
                break;
            }
            if (c == 'd') {
                dateFormat = "dd-MM kk:mm";
                break;
            }
        }
        File logFile = getLogFile(context);
        if (logFile != null) {
            Log.w(TAG, "logging to " + logFile);

            if (logFile.isFile() && logFile.exists()) {
                rotateAsync(logFile);
            }
            try {
                writer = new PrintWriter(new FileWriter(logFile, true));
            } catch (IOException e) {
                Log.w(TAG, "error opening app log", e);
            }
        }
    }

    public void append(String s) {
        if (writer == null) return;

        StringBuilder sb = new StringBuilder();
        sb.append(format(new Date()))
                .append(" ").append(s);
        writer.println(sb);
        if (LOCAL_LOGV) Log.v(TAG, "[AppLog]: " + sb);
    }

    public void appendAndClose(String s) {
        append(s);
        close();
    }

    public void close() {
        if (LOCAL_LOGV) Log.v(TAG, "AppLog#close()");
        if (writer != null) writer.close();
    }

    public CharSequence format(Date d) {
        return DateFormat.format(dateFormat, d);
    }

    public static Dialog displayAsDialog(Context context) {
        File file = getLogFile(context);
        if (file != null && file.exists() && file.length() > 0) {
            return getLogDialog(context, file);
        } else {
            return logNotAvailable(context);
        }
    }

    private static Dialog logNotAvailable(Context context) {
        return new AlertDialog.Builder(context)
            .setTitle(R.string.menu_view_log)
            .setMessage(R.string.app_log_empty)
            .setIcon(android.R.drawable.ic_dialog_info)
            .setPositiveButton(android.R.string.ok, null)
            .create();
    }

    private static Dialog getLogDialog(Context context, @NonNull File file) {
        final int PAD = 5;
        final TextView view = new TextView(context);
        view.setId(ID);
        view.setText(readLog(file));

        final ScrollView scrollView = new ScrollView(context) {
            {
                addView(view);
                setPadding(PAD, PAD, PAD, PAD);
            }

            @Override
            protected void onLayout(boolean changed, int l, int t, int r, int b) {
                super.onLayout(changed, l, t, r, b);
                scrollTo(0, view.getHeight());
            }
        };
        return new AlertDialog.Builder(context)
            .setPositiveButton(android.R.string.ok, null)
            .setTitle(file.getPath())
            .setView(scrollView)
            .create();
    }

    private static String readLog(File f) {
        StringBuilder text = new StringBuilder();
        BufferedReader br = null;
        try {
            br = new BufferedReader(new FileReader(f));
            String line;
            while ((line = br.readLine()) != null) {
                text.append(line);
                text.append('\n');
            }
        } catch (IOException e) {
            Log.e(TAG, "error reading", e);
        } finally {
            if (br != null) {
                try {
                    br.close();
                } catch (IOException ignored) {
                }
            }
        }
        return text.toString();
    }

    private static @Nullable File getLogFile(Context context) {
        final File logDir = context.getExternalFilesDir(null);
        if (logDir != null) {
            return new File(logDir, App.LOG);
        } else {
            return null;
        }
    }

    private void rotateAsync(final File logFile) {
        new Thread() {
            @Override
            public void run() {
                rotate(logFile);
            }
        }.start();
    }

    private void rotate(final File logFile) {
        if (logFile.length() < MAX_SIZE) return;
        if (LOCAL_LOGV) Log.v(TAG, "rotating logfile " + logFile);

        try {
            LineNumberReader r = new LineNumberReader(new FileReader(logFile));

            while (r.readLine() != null) ;
            r.close();

            int keep = Math.round(r.getLineNumber() * 0.3f);
            if (keep > 0) {
                r = new LineNumberReader(new FileReader(logFile));

                while (r.readLine() != null && r.getLineNumber() < keep) ;

                File newFile = new File(logFile.getAbsolutePath() + ".new");
                PrintWriter pw = new PrintWriter(new FileWriter(newFile));
                String line;
                while ((line = r.readLine()) != null) pw.println(line);

                pw.close();
                r.close();

                if (newFile.renameTo(logFile) && LOCAL_LOGV) {
                    Log.v(TAG, "rotated file, new size = " + logFile.length());
                }
            }
        } catch (IOException e) {
            Log.e(TAG, "error rotating file " + logFile, e);
        }
    }
}
