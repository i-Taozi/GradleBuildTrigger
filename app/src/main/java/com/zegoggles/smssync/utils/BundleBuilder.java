package com.zegoggles.smssync.utils;

import android.os.Bundle;
import android.os.Parcelable;

import java.util.ArrayList;

/**
 * A Bundle that doesn't suck. Allows you to chain method calls as you'd expect.
 * https://gist.github.com/felipecsl/45bc010713c0f543354e
 *
 * MIT license
 */
public class BundleBuilder {
    private final Bundle bundle;

    public BundleBuilder() {
        this(new Bundle());
    }

    public BundleBuilder(Bundle existingBundle) {
        bundle = new Bundle(existingBundle);
    }

    public BundleBuilder putBoolean(String key, boolean value) {
        bundle.putBoolean(key, value);
        return this;
    }

    public BundleBuilder putByte(String key, byte value) {
        bundle.putByte(key, value);
        return this;
    }

    public BundleBuilder putChar(String key, char value) {
        bundle.putChar(key, value);
        return this;
    }

    public BundleBuilder putShort(String key, short value) {
        bundle.putShort(key, value);
        return this;
    }

    public BundleBuilder putInt(String key, int value) {
        bundle.putInt(key, value);
        return this;
    }

    public BundleBuilder putLong(String key, long value) {
        bundle.putLong(key, value);
        return this;
    }

    public BundleBuilder putFloat(String key, float value) {
        bundle.putFloat(key, value);
        return this;
    }

    public BundleBuilder putDouble(String key, double value) {
        bundle.putDouble(key, value);
        return this;
    }

    public BundleBuilder putString(String key, String value) {
        bundle.putString(key, value);
        return this;
    }

    public BundleBuilder putParcelable(String key, Parcelable value) {
        bundle.putParcelable(key, value);
        return this;
    }

    public BundleBuilder putParcelableArray(String key, Parcelable[] value) {
        bundle.putParcelableArray(key, value);
        return this;
    }

    public BundleBuilder putParcelableArrayList(String key, ArrayList<? extends Parcelable> value) {
        bundle.putParcelableArrayList(key, value);
        return this;
    }

    public BundleBuilder putAll(Bundle bundle) {
        this.bundle.putAll(bundle);
        return this;
    }

    /**
     * Converts this BundleBuilder into a plain Bundle object.
     *
     * @return A new Bundle containing all the mappings from the current BundleBuilder.
     */
    public Bundle build() {
        return new Bundle(bundle);
    }
}
