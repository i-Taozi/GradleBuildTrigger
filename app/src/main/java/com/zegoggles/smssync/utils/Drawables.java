package com.zegoggles.smssync.utils;

import android.content.res.Resources;
import android.graphics.drawable.Drawable;
import androidx.annotation.NonNull;
import androidx.core.content.res.ResourcesCompat;
import androidx.core.graphics.drawable.DrawableCompat;

public class Drawables {
    private Drawables() {}

    @NonNull
    public static Drawable getTinted(Resources resources, int resource, int color) {
        Drawable drawable = ResourcesCompat.getDrawable(resources, resource, null);
        drawable = DrawableCompat.wrap(drawable);
        DrawableCompat.setTint(drawable.mutate(), color);
        return drawable;
    }
}
