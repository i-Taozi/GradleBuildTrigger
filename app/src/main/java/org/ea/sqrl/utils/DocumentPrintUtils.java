package org.ea.sqrl.utils;

import android.app.Activity;
import android.content.pm.PackageInfo;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.text.Layout;
import android.text.StaticLayout;
import android.text.TextPaint;
import android.util.Log;

import org.ea.sqrl.R;

import java.text.SimpleDateFormat;
import java.util.Date;

/**
 * This class hosts utility functions for drawing print documents.
 *
 * @author Alex Hauser
 */
public class DocumentPrintUtils {
    private static final String TAG = "DocumentPrintUtils";

    public static int drawTextBlock(Canvas canvas, String text, Layout.Alignment alignment, TextPaint textPaint, int y, int margin) {
        int width = canvas.getWidth() - (margin * 2);

        StaticLayout staticLayout = new StaticLayout(
                text,
                textPaint,
                width,
                alignment,
                1,
                0,
                false);

        canvas.save();
        canvas.translate(margin, y);
        staticLayout.draw(canvas);
        canvas.restore();

        return staticLayout.getHeight();
    }

    public static void drawPrintPageFooter(Activity activity, Canvas canvas) {
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");

        StringBuilder versionString = new StringBuilder();
        versionString.append(activity.getString(R.string.print_version_string));
        versionString.append(" ");
        versionString.append(sdf.format(new Date()));
        versionString.append(" * ");
        versionString.append("Version: ");

        Bitmap loadLogo = BitmapFactory.decodeResource(activity.getResources(), R.drawable.sqrl_print_logo);
        Bitmap sqrlLogo = Bitmap.createScaledBitmap(loadLogo, 40, 40, false);

        TextPaint textPaint = new TextPaint();
        textPaint.setAntiAlias(true);
        textPaint.setTextSize(10);
        textPaint.setFakeBoldText(false);
        textPaint.setColor(Color.DKGRAY);

        canvas.drawBitmap(
                sqrlLogo,
                (canvas.getWidth()/2) - (sqrlLogo.getScaledWidth(canvas) / 2),
                canvas.getHeight() - 140,
                new Paint()
        );

        try {
            PackageInfo pInfo = activity.getPackageManager().getPackageInfo(activity.getPackageName(), 0);
            versionString.append(pInfo.versionName);
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
        }

        drawTextBlock(
                canvas,
                versionString.toString(),
                Layout.Alignment.ALIGN_CENTER,
                textPaint,
                canvas.getHeight() - 80,
                20);

        drawTextBlock(canvas,
                "https://github.com/kalaspuffar/secure-quick-reliable-login",
                Layout.Alignment.ALIGN_CENTER,
                textPaint,
                canvas.getHeight() - 65,
                20);
    }
}
