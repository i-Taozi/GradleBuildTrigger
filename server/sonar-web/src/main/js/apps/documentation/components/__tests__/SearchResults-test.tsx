/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import lunr, { LunrToken } from 'lunr';
import * as React from 'react';
import { mockDocumentationEntry } from '../../../../helpers/testMocks';
import { getUrlsList } from '../../navTreeUtils';
import { DocumentationEntry } from '../../utils';
import SearchResultEntry from '../SearchResultEntry';
import SearchResults, { tokenContextPlugin, tokenContextPluginCallback } from '../SearchResults';

jest.mock('../../navTreeUtils', () => ({
  getUrlsList: jest.fn().mockReturnValue([])
}));

jest.mock('lunr', () =>
  jest.fn(() => ({
    search: jest.fn(() => [
      {
        ref: 'lorem/origin',
        matchData: {
          metadata: {
            simply: {
              title: { position: [[19, 5]] },
              text: {
                position: [
                  [15, 6],
                  [28, 4]
                ],
                tokenContext: ['is simply dummy', 'simply dummy text']
              }
            }
          }
        }
      },
      {
        ref: 'foobar',
        matchData: {
          metadata: {
            simply: {
              title: { position: [[23, 4]] },
              text: {
                position: [
                  [111, 6],
                  [118, 4]
                ],
                tokenContext: ['dummy simply text']
              }
            }
          }
        }
      }
    ])
  }))
);

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot('default');
  expect(shallowRender({ query: '' }).type()).toBeNull();
});

describe('search engine', () => {
  class LunrIndexMock {
    plugins: Function[] = [];
    fields: Array<{ field: string; args: any }> = [];
    docs: DocumentationEntry[] = [];
    metadataWhitelist: string[] = [];

    use(fn: Function) {
      this.plugins.push(fn);
    }

    ref(_ref: string) {
      /* noop */
    }

    field(field: string, args = {}) {
      this.fields.push({ field, args });
    }

    add(doc: DocumentationEntry) {
      this.docs.push(doc);
    }
  }

  it('should correctly populate the index', () => {
    (getUrlsList as jest.Mock).mockReturnValueOnce(['/lorem/index', '/lorem/origin']);

    shallowRender();

    // Fetch the callback passed to lunr(), which serves as the index constructor.
    const indexConstructor: Function = (lunr as jest.Mock).mock.calls[0][0];

    // Apply it to our mock index.
    const lunrMock = new LunrIndexMock();
    indexConstructor.apply(lunrMock);

    expect(lunrMock.docs.length).toBe(2);
    expect(lunrMock.plugins).toContain(tokenContextPlugin);
    expect(lunrMock.metadataWhitelist).toEqual(['position', 'tokenContext']);
    expect(lunrMock.fields).toEqual([
      expect.objectContaining({ field: 'title', args: { boost: 10 } }),
      expect.objectContaining({ field: 'text' })
    ]);
  });

  it('should correctly look for an exact match', () => {
    // No exact match, should sort as the matches came in.
    const wrapper = shallowRender({ query: 'text simply' });
    expect(
      wrapper
        .find(SearchResultEntry)
        .at(0)
        .props().result.page.relativeName
    ).toBe('lorem/origin');

    // Exact match, specific page should be at the top.
    wrapper.setProps({ query: 'simply text' });
    expect(
      wrapper
        .find(SearchResultEntry)
        .at(0)
        .props().result.page.relativeName
    ).toBe('foobar');
  });

  it('should trigger a search if query is set', () => {
    const wrapper = shallowRender({ query: undefined });
    expect(wrapper.instance().index.search).not.toBeCalled();
    wrapper.setProps({ query: 'si:+mply text' });
    expect(wrapper.instance().index.search).toBeCalledWith('simply~1 simply* text~1 text*');
  });
});

describe('tokenContextPluginCallback', () => {
  class LunrTokenMock {
    str: string;
    metadata: any;

    constructor(str: string) {
      this.str = str;
      this.metadata = {};
    }

    toString() {
      return this.str;
    }
  }

  function mockLunrToken(str: string): LunrToken {
    return new LunrTokenMock(str);
  }

  it('should correctly provide token context for text', () => {
    const tokens = [
      mockLunrToken('this'),
      mockLunrToken('is'),
      mockLunrToken('some'),
      mockLunrToken('text')
    ];

    expect(tokenContextPluginCallback(mockLunrToken('this'), 0, tokens).metadata).toEqual(
      expect.objectContaining({ tokenContext: 'this is' })
    );
    expect(tokenContextPluginCallback(mockLunrToken('is'), 1, tokens).metadata).toEqual(
      expect.objectContaining({ tokenContext: 'this is some' })
    );
    expect(tokenContextPluginCallback(mockLunrToken('some'), 2, tokens).metadata).toEqual(
      expect.objectContaining({ tokenContext: 'is some text' })
    );
    expect(tokenContextPluginCallback(mockLunrToken('text'), 3, tokens).metadata).toEqual(
      expect.objectContaining({ tokenContext: 'some text' })
    );
  });
});

function shallowRender(props: Partial<SearchResults['props']> = {}) {
  return shallow<SearchResults>(
    <SearchResults
      navigation={['lorem/index', 'lorem/origin', 'foobar']}
      pages={[
        mockDocumentationEntry({
          title: 'Lorem Ipsum',
          relativeName: 'lorem/index',
          url: '/lorem/index',
          content:
            "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.",
          text:
            "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book."
        }),
        mockDocumentationEntry({
          title: 'Where does it come from?',
          relativeName: 'lorem/origin',
          url: '/lorem/origin',
          content:
            'Contrary to popular belief, Lorem Ipsum is not simply random text. It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years old. Richard McClintock, a Latin professor at Hampden-Sydney College in Virginia, looked up one of the more obscure Latin words.',
          text:
            'Contrary to popular belief, Lorem Ipsum is not simply random text. It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years old. Richard McClintock, a Latin professor at Hampden-Sydney College in Virginia, looked up one of the more obscure Latin words.'
        }),
        mockDocumentationEntry({
          title: 'Where does Foobar come from?',
          relativeName: 'foobar',
          url: '/foobar',
          content:
            'Foobar is a universal variable understood to represent whatever is being discussed. Now we need some keywords: simply text.',
          text:
            'Foobar is a universal variable understood to represent whatever is being discussed. Now we need some keywords: simply text.'
        })
      ]}
      query="what is 42"
      splat="foobar"
      {...props}
    />
  );
}
