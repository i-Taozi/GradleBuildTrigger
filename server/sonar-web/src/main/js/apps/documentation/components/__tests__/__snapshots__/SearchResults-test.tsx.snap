// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly: default 1`] = `
<Fragment>
  <SearchResultEntry
    active={false}
    key="lorem/origin"
    result={
      Object {
        "exactMatch": false,
        "highlights": Object {
          "text": Array [
            Array [
              15,
              6,
            ],
            Array [
              28,
              4,
            ],
          ],
          "title": Array [
            Array [
              19,
              5,
            ],
          ],
        },
        "longestTerm": "",
        "page": Object {
          "content": "Contrary to popular belief, Lorem Ipsum is not simply random text. It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years old. Richard McClintock, a Latin professor at Hampden-Sydney College in Virginia, looked up one of the more obscure Latin words.",
          "navTitle": undefined,
          "relativeName": "lorem/origin",
          "text": "Contrary to popular belief, Lorem Ipsum is not simply random text. It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years old. Richard McClintock, a Latin professor at Hampden-Sydney College in Virginia, looked up one of the more obscure Latin words.",
          "title": "Where does it come from?",
          "url": "/lorem/origin",
        },
        "query": "what is 42",
      }
    }
  />
  <SearchResultEntry
    active={true}
    key="foobar"
    result={
      Object {
        "exactMatch": false,
        "highlights": Object {
          "text": Array [
            Array [
              111,
              6,
            ],
            Array [
              118,
              4,
            ],
          ],
          "title": Array [
            Array [
              23,
              4,
            ],
          ],
        },
        "longestTerm": "",
        "page": Object {
          "content": "Foobar is a universal variable understood to represent whatever is being discussed. Now we need some keywords: simply text.",
          "navTitle": undefined,
          "relativeName": "foobar",
          "text": "Foobar is a universal variable understood to represent whatever is being discussed. Now we need some keywords: simply text.",
          "title": "Where does Foobar come from?",
          "url": "/foobar",
        },
        "query": "what is 42",
      }
    }
  />
</Fragment>
`;
